GH-96803: Document and test new unstable internal frame API functions (#104211)
python/cpython: Doc/c-api/frame.rst

.. c:function:: int PyFrame_GetLineNumber(PyFrameObject *frame)

   Return the line number that *frame* is currently executing.

Internal Frames
---------------

> Unless using PEP 523, you will not need this.

Reviewer: ```suggestion
Unless using :pep:`523`, you will not need this.
```

.. c:struct:: _PyInterpreterFrame

   The interpreter's internal frame representation.

> 

Reviewer: `_PyInterpreterFrame` was added in Python 3.11, right?

```suggestion

   .. versionadded:: 3.11
```

.. c:function:: PyCodeObject* PyUnstable_InterpreterFrame_GetCode(struct _PyInterpreterFrame *frame);

    Return a :term:`strong reference` to the code object for the frame.

   .. versionadded:: 3.12

.. c:function:: int PyUnstable_InterpreterFrame_GetLasti(struct _PyInterpreterFrame *frame);

   Return the byte offset into the last executed instruction.

   .. versionadded:: 3.12

.. c:function:: int PyUnstable_InterpreterFrame_GetLine(struct _PyInterpreterFrame *frame);

>    Return the  currently executing line number, or -1 if there is no line number.

Reviewer 2: ```suggestion
   Return the currently executing line number, or -1 if there is no line number.
```

.. c:function:: int PyFrame_GetLineNumber(PyFrameObject *frame)

   Return the line number that *frame* is currently executing.

Internal Frames
---------------

Unless using PEP 523, you will not need this.

> .. c:type:: struct _PyInterpreterFrame

Reviewer: ```suggestion
.. c:type:: _PyInterpreterFrame
```

Author: The header doesn't include a `typedef`, so I think `struct _PyInterpreterFrame` is correct. Unless we always skip `struct` in the docs for some reason.

Reviewer: `_PyInterpreterFrame` is defined in `pycore_frame.h` with `typedef struct _PyInterpreterFrame { ... } _PyInterpreterFrame`.

It looks like we are inconsistent in code about referring to it as `_PyInterpreterFrame*` vs `struct _PyInterpreterFrame*`, but there are plenty of examples of both. Maybe it originally was defined without the `typedef`, but that was changed at some point?

In any case, there aren't any other examples in the docs of using `c:type::` with `struct ...`, and Sphinx seems to not like it *shrug*.

Reviewer: It looks like if you prefer to document this as a struct rather than a type, the way to do that would be to use `c:struct::` instead of `c:type::` (and don't explicitly include the `struct` keyword, the `c:struct::` will add it for you in the rendered docs.) See for example `PyCompilerFlags` in `Doc/c-api/veryhigh.rst`.

Reviewer 3: Carl is right, `.. c:struct::` should be preferred. See the docs for reference: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#the-c-domain

A

>    This C structure for internal frames.

Reviewer: ```suggestion
   The interpreter's internal frame representation.
```

.. c:function:: PyCodeObject* PyUnstable_InterpreterFrame_GetCode(struct _PyInterpreterFrame *frame);

>     Get the code object for the frame
>     Returns a :term:`strong reference`.

Reviewer: ```suggestion
    Return a :term:`strong reference` to the code object for the frame.
```

   .. versionadded:: 3.12

.. c:function:: int PyUnstable_InterpreterFrame_GetLasti(struct _PyInterpreterFrame *frame);

>    Gets the byte offset into the last executed instruction.

Reviewer: ```suggestion
   Return the byte offset into the last executed instruction.
```

   .. versionadded:: 3.12

.. c:function:: int PyUnstable_InterpreterFrame_GetLine(struct _PyInterpreterFrame *frame);

>    Gets the  currently executing line number, or -1 if there is no line number.

Reviewer: ```suggestion
   Return the  currently executing line number, or -1 if there is no line number.
```
